[VideoVersions] Revamp Navigation to Versions/Extras and Remove Selection Dialog

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

This is a reimplementation of the GUI navigation of movie versions and extras for better consistency with the rest of Kodi.

The GUI is an extension of the "Multiple versions as folder" setting: movies with multiple versions now act like folders that contain the versions as well as a virtual subfolder for extras when applicable.
The version/extra selection dialog was removed, as well as a few context menu entries.
The manager dialog remains as is.
Removed the "multiple versions as folder" setting since it's the only way now.
Navigation from top level versions video node is unchanged.

The most significant change from a code point of view is the creation of new node types to extend videodb:// library urls to represent unambiguously versions and extras.

existing:
videodb://movies/titles/0 > all movies
videodb://movies/titles/23 > a specific movie
videodb://movies/videoversions/x/y with x=version type id (name) and y=movie id - cannot describe extras and ambiguous for movies that have multiple versions of same name

new:
videodb://movies/titles/23/x returns a list of the movie's assets of asset type x (column itemType of the videoversion table)
x = -2 > special value for all versions of a movie plus a virtual folder Extras if applicable for the movie
x = 0 > special value for all assets of a movie (versions and extras)
x = 1,2 > assets of the given type for the movie (1= all versions, 2= all extras)

videodb://movies/titles/23/x/y describes a specific asset
x - same meaning as above
y - file id (unique identifier) of the video asset

I have limited understanding of the underlying nodes navigation and its capabilties. The extension above is adhoc for movies and may not extend well to tv shows/music videos/others in the future.
An alternative could be like videodb://assets/[mediatype]/[mediaid]/[assettype]/[assetid = idFile]. I also don't know enough about how that intersects with possible future query requirements, like retrieving only a specific resolution, codec, quality, ... without resorting to post query filtering instead of backend db filtering.
Would be nice to fold the current Versions root node navigation of the different editions into a more generic framework.

Skinning changes:
- ~~new infobool ListItem.IsVideoAsset to let skins add the versions/extras icons only for movie items and not for specific versions or for extras.~~ not necessary anymore because of new content values
- new items list content values videoassets (possible mix of versions and assets), videoversions and videoextras
 
Known issues:
- ~~need to make the list of versions look a bit like the former selection dialog with very visible media flags (help choose between different qualities of the same movie edition for example)~~ new MediaList view thanks to jjd-uk
- ~~the default version is not selected by default when displaying the versions~~
- ~~play in context menu plays default version for movie with multiple versions but plays all extras for extras folder~~ it's a feature, not a bug!
- ~~a favorited movie with multiple versions failed to open the assets list when selected~~
- ~~need to adapt the fix of PR 24645 rather than simply rip it out, the issue remains for the hybrid folder items~~
- examine all code locations with @todo get rid of \"videos with versions as folder\" hack! comment
- list of versions/extras not refreshed after info dialog + change with manager dialog like rename version / change art / ...
- blurry icon on extras virtual folder, need a larger image in Estuary - or use a standard folder icon.
- ~~which skin views should be enabled for assets?~~

Possible future improvements:
- setting to decide action when for play item of context menu on a movie pseudo-folder (play/queue default version, all versions, default version + extras, ...)
- re-enable and adapt standard context menu  actions on versions and extras like set title, set art, remove
- what to show in info dialog for versions and extras - they don't have own metadata
- ~~navigation labels at the top left of screen for Estuary: would be nice to add more components to the navigation path (currently 2 levels only. ex. Movies / Big Buck Bunny, might be nice to be able to do Movies / Versions / Big Buck Bunny)~~
- setting to control which version is autoselected on first entry of the versions virtual folder. None, default version, first unwatched, first in-progress?

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Modify the native versions and extras feature to fit better with Kodi's architecture and GUI, start removing temporary workarounds added to get v21 released.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Exercised most code paths, added some unit tests.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Modified GUI for versions and extras, now more consistent with the rest of Kodi.

## Screenshots (if appropriate):
Movies list: no change
![image](https://github.com/user-attachments/assets/e6af569a-09d2-45dc-be23-a7d8c0f688cc)

After selecting a movie that has multiple versions and/or extras (List view):
![image](https://github.com/user-attachments/assets/d605215f-4bf6-4787-94b5-d1866103f383)

After selecting a movie that has multiple versions and/or extras (new MediaList view):
![image](https://github.com/user-attachments/assets/d3dc015b-2969-4a0d-9f1d-010bf0142798)
(new MediaList view)

After selecting the extras subfolder:
![image](https://github.com/user-attachments/assets/531d34ab-5864-4045-92d6-e09443c7ce22)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

CrystalP committed:
[videodb] Renumber video version itemType, bump DB

CrystalP committed:
[video] Integrate movie versions and extras in directory nodes and normal navigation

CrystalP committed:
[video] Remove Choose Version / Play Version Using from context menu

CrystalP committed:
[Estuary] Don't show versions/extras icons for video assets

CrystalP committed:
[videoversions] Remove "versions as folder" setting and remaining related code

CrystalP committed:
[guilib] Remove Choose Version/Extras Dialog and related functions

CrystalP committed:
[Estuary] Remove support of the Version/Extras select dialog

CrystalP commented:
no changes in that force push, just fixed clang formatting and removed some commented out code.

CrystalP commented:
one more time to shut up clang formatting warnings

CrystalP commented:
win x64 build kindly kicked off by hitcher
https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20250116-80fc733d-feat-videoassets-nav-x64.exe

garbear commented on addons/skin.estuary/xml/Variables.xml:
```diff
@@ -850,9 +850,8 @@
-		<value condition="Window.IsVisible(selectvideoversion)">$INFO[ListItem.VideoVersionName]</value>
 		<value>$INFO[ListItem.Label]</value>
-	</variable>
+	</variable>	
```
Trailing whitespace

garbear commented:
Nice, I like the asset-centric approach. I can tell it's more generic because of how much code this PR removes. I haven't runtime tested but a quick look at the code looks good.

CrystalP committed:
[video] Select the default version on first entry of a movie assets collection

CrystalP commented:
- removed missed whitespace, removed a bit more now unused old code
- additional commit to select the default movie version by default. Could be worth a setting if some users prefer to remain on the .. parent item.

CrystalP committed:
[Estuary] Include the asset type in the TopBarLabels for asset content

Hitcher commented:
> * additional commit to select the movie default version by default in the versions list. Could be worth a setting if some users prefer to remain on the .. parent item.

Or just the first item because the parent item can be hidden.

CrystalP commented:
> Or just the first item because the parent item can be hidden.

I didn't understand? Do you mean if .. is removed? First item is always by default afaik.
Thought of another possible preference: auto select the first in-progress version.

Hitcher commented:
> Do you mean if .. is removed?

Yes, when disabling "Show parent folder items".

CrystalP committed:
[video] Separate content values for mixed video assets, video versions and video extras.

CrystalP commented:
Build with content videoasssets videoversions videoextras for the skin to distinguish from regular movies listing
https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20250118-6f610641--x64.exe

CrystalP committed:
[Estuary] Make Additional Views Available for videoassets/versions/extras Content

CrystalP committed:
[video] Pass asset id from videodb paths to GetMovieInfo()

Finish the correction started by PR#26590. The asset id (=file id) is available through CQueryParams now.

CrystalP commented:
Added some unit tests for the new node types.
Added a fix for the part of #26590 that couldn't be done without the navigation redesign.

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -6669,11 +6669,40 @@ void CVideoDatabase::UpdateTables(int iVersion)
+    // former value 0 for versions becomes 1
+    // former value 1 for extras becomes 2
+
+    constexpr int VIDEOASSETTYPE_VERSION_OLD = 0;
+    constexpr int VIDEOASSETTYPE_EXTRA_OLD = 1;
```
`static constexpr` - you don't want the constants created on every function invocation.

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssetTypes.h:
```diff
@@ -0,0 +1,26 @@
+
+#include "DirectoryNode.h"
+
+namespace XFILE
+{
+namespace VIDEODATABASEDIRECTORY
```
To save some lines of code you could write 

```c++
namespace XFILE::VIDEODATABASEDIRECTORY
{
  ..
```

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssets.cpp:
```diff
@@ -0,0 +1,67 @@
+bool CDirectoryNodeMovieAssets::GetContent(CFileItemList& items) const
+{
+  CVideoDatabase videodatabase;
+  if (!videodatabase.Open())
+    return false;
```
Log an error?

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssets.cpp:
```diff
@@ -0,0 +1,67 @@
+
+  videodatabase.Close();
+
+  if (params.GetVideoAssetType() == -2 && items.Size() && items[0]->HasVideoExtras() &&
```
Could we introduce a constant with a speaking name to get rid of the magic number `-2`?

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssets.cpp:
```diff
@@ -0,0 +1,67 @@
+    const std::string pathExtras{URIUtils::AddFileToFolder(
+        URIUtils::GetParentPath(path), std::to_string(static_cast<int>(VideoAssetType::EXTRA)))};
+
+    CFileItemPtr pItem(new CFileItem(pathExtras, true));
```
`const auto item{std::make_shared<CFileItem>(pathExtras, ...`

Please also remove the `p`prefix. Coding guidelines.

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssets.h:
```diff
@@ -0,0 +1,26 @@
+
+#include "DirectoryNode.h"
+
+namespace XFILE
+{
+namespace VIDEODATABASEDIRECTORY
```
see previous comment, could be shortened.

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/QueryParams.h:
```diff
@@ -60,6 +62,8 @@ namespace XFILE
     long m_idSet;
     long m_idTag;
     long m_idVideoVersion{-1};
+    long m_videoAssetType{-1};
+    long m_idVideoAsset{-1};
```
Just asking: in the mood to get rid of the magic numbers (which you have not introduced - I know)?

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/test/TestNodeMovieAssetTypes.cpp:
```diff
@@ -0,0 +1,42 @@
+
+TEST(TestNodeMovieAssetTypes, General)
+{
+  std::unique_ptr<CDirectoryNode> pNode(
```
Remove the type prefix from the variable's name, please. Here, and everywhere else in this PR where you introduce new variables.

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -8560,26 +8550,42 @@ bool CVideoDatabase::GetMoviesByWhere(const std::string& strBaseDir, const Filte
+          {
+            pItem->SetProperty("IsHybridFolder", true);
+            pItem->m_bIsFolder = true;
+            itemUrl.AppendPath("-2/");
```
Magic number...

ksooo commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -7278,6 +7278,14 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
+///   \table_row3{   <b>`ListItem.IsVideoAsset`</b>,
+///                  \anchor ListItem_IsVideoAsset
+///                  _boolean_,
+///     @return **True** when the selected item is a video asset (for example, a specific movie version or an extra).
```
comma needs to be escaped. => `example\, a`

ksooo commented on xbmc/guilib/guiinfo/VideoGUIInfo.cpp:
```diff
@@ -840,6 +840,10 @@ bool CVideoGUIInfo::GetBool(bool& value, const CGUIListItem *gitem, int contextW
         value = tag->HasVideoExtras();
         return true;
+      case LISTITEM_ISVIDEOASSET:
+        //! @todo: is this a hot path? should store in the tag instead of recalculating?
+        value = VIDEO::IsVideoAssetFile(*item);
+        return true;
```
Potentially every gui label/bool/int is `hot` as you never know what skinners will do with it. So, I strongly suggest to store the value at the tag.

CrystalP replied:
After investigating, it doesn't lend itself well to a new field in the tag as it's not stored with the file but depends on the context.

IsVideoAssetFile is not that expensive, but there is parsing of the path to a CURL. I don't know where the bar for "acceptable" is.

One alternative would be a new property or a new member of CFileItem (where is the bar?),set by GetMoviesByWhere.

Other alternative (commit added to the PR - would replace the earlier commits for a merge) is to skip the new infobool and to have the skin use Container.Content(videoassets) | Container.Content(videoversions) | Container.Content(videoExtras) in conditions instead. Same results with the current skin/core usage.
I didn't know how to create a variable to avoid repeating the condition in multiple places in Estuary, to be improved by a skinner.

Noticed a pattern of OR'ing many Container.Content(x) statements in a few places of Estuary, like Container.Content(x)|Container.Content(y)|Container.Content(z)|... It's likely similar in other skins and can't be very good for performance. I may contribute later an improvement to Container.Content() to accept | separated strings, like Container.Content(x|y|z)

ksooo commented on xbmc/video/windows/GUIWindowVideoNav.h:
```diff
@@ -62,5 +62,6 @@ class CGUIWindowVideoNav : public CGUIWindowVideoBase
   virtual IncludeAllSeasonsAndSpecials GetSettingIncludeAllSeasonsAndSpecials();
   virtual int GetFirstUnwatchedItemIndex(bool includeAllSeasons, bool includeSpecials);
-  void SelectFirstUnwatched();
+  int GetFirstSelectedItemIndex();
```
could this method be `const`?

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -8496,7 +8496,13 @@ bool CVideoDatabase::GetMoviesByWhere(const std::string& strBaseDir, const Filte
     // navigation = list of assets of the movie
-    const bool assetsNav{options.find("assetType") != options.end()};
+    bool assetsNav{false};
+    int assetType{-1};
```
just saying: magic number

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -8558,7 +8564,27 @@ bool CVideoDatabase::GetMoviesByWhere(const std::string& strBaseDir, const Filte
           if (!items.HasProperty("customtitle"))
-            items.SetProperty("customtitle", movie.GetTitle());
+          {
+            int msgId;
```
I know not actually needed here, but for future safety: defualt init?

CrystalP replied:
I raise you a lambda to make it const int :-)

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssetTypes.cpp:
```diff
@@ -18,5 +21,18 @@ CDirectoryNodeMovieAssetTypes::CDirectoryNodeMovieAssetTypes(const std::string&
+
+  switch (params.GetVideoAssetType())
+  {
+    case -2: // special value for all assets + extras virtual folder
```
magic number ... and you even wrote a comment what it represents (iirc, not for the first time in this PR) ;-)

ksooo reviewed (commented):
PR was very easy to review, because very well separated into different small commits.

So nice to see those hacks go we introduced under heavy pressure back in the days we needed to make the feature usable at all.

CrystalP commented:
Thanks for the review.
-2 magic number: I'll think about it. It's likely to create some coupling - or that could be a sign of things not properly abstracted and that should not be coupled..
Ideally there would be one place explaining all nodes, I didn't find any doc.

CrystalP commented:
Addressed the comments, still thinking about the magic numbers.
See answer about the ListItem.IsVideoAsset InfoBool, commit added with other solution. Would replace the original commits.

CrystalP commented:
Fixed the formatting, replaced the -2 magic number with a new VideoAssetType value, VERSIONSANDEXTRASFOLDER and removed a bit more now unused code.

ksooo reviewed (approved):
Code changes look good to me.
 
Cannot say anything about the ListItem.IsVideoAsset InfoBool topic. This is more for a skinner to comment.

ksooo commented on xbmc/video/VideoDbUrl.cpp:
```diff
@@ -190,6 +195,10 @@ bool CVideoDbUrl::parse()
     AddOption("year", (int)queryParams.GetYear());
   if (queryParams.GetVideoVersionId() != -1)
     AddOption("videoversionid", (int)queryParams.GetVideoVersionId());
+  if (queryParams.GetVideoAssetType() != -1)
+    AddOption("assetType", (int)queryParams.GetVideoAssetType());
+  if (queryParams.GetVideoAssetId() != -1)
+    AddOption("assetid", (int)queryParams.GetVideoAssetId());
```
Replace the c-style cast with c++-style cast?

jjd-uk commented:
> Code changes look good to me.
> 
> Cannot say anything about the ListItem.IsVideoAsset InfoBool topic. This is more for a skinner to comment.

Sorry but I'm not sure what exactly needs to be commented on?

ksooo commented:
There is no new info bool anymore. The last commit in this PR, as a functional alternative, using already existing bools.

CrystalP commented:
Replaced the last commit since I missed a couple lines in Variables.xml.

@jjd-uk  Since there are now specific content values for assets, I suggest to use them to decide whether to show the versions/extras overlays.
See Variables.xml, View_51_Poster.xml, View_53_Shift.xml and View_54_InfoWall

However if for reasons there must be a per-item infobool like IsVideoAsset, then I thought about it a bit again and saving CVideoDbUrl or CQueryParams in the videotag would provide some flexibility beyond the immediate need with an access to some context info of the item in core, and would also allow a more efficient VideoFileItemClassify::IsVideoAssetFile() used in other places in core.

jjd-uk commented:
Ok, I don't see any downside doing it the `ListItem.HasVideoVersions + !Container.Content(videoassets) + !Container.Content(videoversions) + !Container.Content(videoextras)` way.

So yeah I'd agree for `ListItem.IsVideoAsset` have we have the Container.Content items.

Any thoughts @Hitcher?

CrystalP commented:
Squashed a few commits, added a bit more doxygen (Container.Content), added a couple unit tests for VideoDbUrl.
Removed the 2 commits related to sort methods, which raised some questions better handled in a separate PR to not delay this one I think (and there might be some bikeshedding...).

@jjd-uk would you do a version of the MediaList view without workaround for missing label2 on some sorting methods? I could cherry-pick it here.

jjd-uk committed:
[Estuary] Add new view and adjust views for videoassets/versions/extras

CrystalP commented:
Fixed the formatting comment, included jjd-uk's initial version of MediaList view.

As mentioned in Slack, the concatenation of the top bar label name should be done in the skin instead of core to give skins more flexibility. See the modified commit "[Estuary] Include the asset type in the TopBarLabels for asset content"
Alterntatively an infobool with the translated asset type could be created so the skin only has to assemble the pieces. Could be changed in next PR.

CrystalP commented:
Fixed the formatting, removed a couple missed item->SetProperty("has_resolved_video_asset", true);
Will merge on green.
